src/NHibernate/Type/GenericOrderedSetType.cs

```diff
@@ -24,9 +27,41 @@ public GenericOrderedSetType(string role, string propertyRef)
 		{
 		}
 
+		static ConstructorInfo SetConstructor
+		{
+			get
+			{
+				if (_setConstructor == null)
+				{
+					lock (typeof(GenericOrderedSetType<>))
```

Comment:
And this `MethodImpl(Synchronized)` is not recommended on public types too by [Msdn](https://msdn.microsoft.com/en-us/library/system.runtime.compilerservices.methodimploptions(v=vs.110).aspx). 

> Locking on the instance or on the type, as with the Synchronized flag, is not recommended for public types, because code other than your own can take locks on public types and instances. This might cause deadlocks or other synchronization problems.

Of course it is unlikely that other code do that. But that the path toward unrecoverable freezes. And still, why not using the `Lazy<T>` class for lazy initialisation where possible?